### DISCUSSION TITLE
How to get the current path of a neo-tree buffer?

### DISCUSSION THREAD

**Discussion by wookayin** [nvim-neo-tree/neo-tree.nvim#319] [Q&A] [answered]
I would like to get the path of an existing neo-tree buffer (i.e., the path of the root node) programatically. Or any node/item the current cursor is pointing to?  The path to the root node is main concern; FYI nerd-tree allows to do this like `b:NERDTree.root.path._str()`.

I searched the documentation for "current path" or looked into `b:` variables but I was unable to find them quickly. Any help, hints, or documentation?

**Comment by cseickel** [comment, marked as answer]
If you use the `bind_to_cwd` option, which is true by default, then a sidebar tree's root will be the same as the tab working directory. A tree opened in a window with the "current" position will use the local working directory of that window.

Alternatively, you can read the state of a tree with:

```lua
local state = require("neo-tree.sources.manager").get_state("filesystem")
print(state.path)
```

That will give you the state of the "filesystem" sidebar in the current tab. The parameters to the `get_state` function are: `source_name, tabnr, winid`

If you want to get the state for a particular window that is not a sidebar, pass the window id:
```lua
local winid = vim.api.nvim_get_current_win()
local state = require("neo-tree.sources.manager").get_state("filesystem", nil, winid)
print(state.path)
```

**Reply by rickyelopez** [reply]
Does the table that is returned from this call include the currently highlighted node anywhere? I'm trying to use this to figure out what is currently selected, but it seems like the `path` member contains a parent node of the currently selected item, and I'm not seeing anything obvious that gets updated on selection change. Is there a function available that I could use to check that?

**Reply by ryanmsnyder** [reply]
I'm also trying to figure out how to get the `path` of the item under the cursor but `require("neo-tree.sources.manager").get_state("filesystem")` only returns the parent node's path. @rickyelopez did you figure out a way to get the item under the cursor in the neo-tree sidebar?

**Reply by cseickel** [reply]
@ryanmsnyder, that method returns the state of the tree, and once you have that you can use it to get the currently focused node:

``` lua
local state = require("neo-tree.sources.manager").get_state("filesystem")
local node = state.tree:get_node()
print(node.path)
```

**Reply by rickyelopez** [reply]
Oh nice that's super useful! Maybe I can revive the project I had in mind for this lol
Thanks @cseickel!

**Reply by ryanmsnyder** [reply]
Thanks @cseickel. I'm using it to display info in my statusline about the current file/directory under my cursor in neo-tree. When I toggle neo-tree on the first time, `get_node()` works perfectly. If I then close neo-tree and then re-open it, I get an error:

```
E5108: Error executing lua .../.local/share/neovim/lazy/nui.nvim/lua/nui/tree/init.lua:279: Invalid 'window': Expected Lua
number
stack traceback:
        [C]: in function 'nvim_win_get_cursor'
        .../.local/share/neovim/lazy/nui.nvim/lua/nui/tree/init.lua:279: in function 'get_node'
```

After some debugging, it appears `nui`'s `get_node` function is passing the previous neo-tree `bufnr` to `vim.fn.win_findbuf(bufnr)`. Since the previous `bufnr` no longer exists, it's not able to find the correct window id of the buffer. 

As an example, when I start up neovim and toggle neo-tree on, the `bufnr` is `15` and I don't get the error - everything works as expected. When I toggle neo-tree off and then back on, I get the error because `nui` is passing a `bufnr` of `15` to `vim.fn.win_findbuf(bufnr)`, which now doesn't exist because neo-tree's `bufnr` is now `16`.

Any idea why this error doesn't occur on the first toggling on of neo-tree but then occurs on subsequent toggles of neo-tree?

**Reply by cseickel** [reply]
My guess is that you have captured an old tree variable and are trying to reuse it later instead of getting the current one. Show me the code and I'll tell you what is wrong.

**Reply by ryanmsnyder** [reply]
I was actually able to ignore the error with a protected call because even though it throws an error, it still works. Like I mentioned above, on the 2nd time (or more) of toggling `neo-tree` on, it calls `vim.fn.win_findbuf(bufnr)` with the old `bufnr` once. What I didn't mention though is that it immediately calls it again right after with the correct `bufnr`.

Here's my code. I'm using [`heirline`](https://github.com/rebelot/heirline.nvim) for my statusline. The below is a component in my statusline:

```lua
local FileSize = {
  provider = function()
    local neo_tree_utils = require "neo-tree.utils"
    local state = require("neo-tree.sources.manager").get_state "filesystem"

    if state.tree ~= nil then
      local success, node = pcall(function()
        return state.tree:get_node() -- where the error occurs 🛑
      end)

      if success then
        local text = "-"
        if node.type == "file" then
          local stat = neo_tree_utils.get_stat(node)
          local size = stat and stat.size or nil
          if size then
            local size_success, human = pcall(neo_tree_utils.human_size, size)
            if size_success then
              text = human or text
            end
          end

          return text
          -- return node.path
        end
      end
    end
  end,
}
```

The above component is a child component of the parent component below, which is only active if the filetype is of type `neo-tree`:

```lua
local NeoTreeStatusLine = {
  condition = function()
    return conditions.buffer_matches {
      filetype = { "neo%-tree" },
    }
  end,

  ViMode,
  Space,
  FileSize,
  Align,
  WorkDirNameBlock,
}
```

Let me know if you need to see more code. I would rather fix the error than catch it but I suppose it's not urgent since it still works as expected. Thanks for your help.

**Reply by cseickel** [reply]
The problem there @ryanmsnyder is that the status line refreshes constantly, so you are actually catching it in the middle of being rendered when there is a tree but not yet a valid buffer. The pcall is fine, but if you want to be more precise there is a method that will tell you if the window is actually fully ready to be interacted with:

```lua
local ready = require("neo-tree.ui.renderer").window_exists(state)
```

**Reply by ryanmsnyder** [reply]
Thanks for the response. I checked if `ready == true` right before calling `state.tree:get_node()` but I'm still getting the same error.

**Comment by wookayin** [comment]
Thanks! I find the latter one more general approach that subsumes the former one, but the latter doesn't work with the neotree opened in the left panel.

More specifically, when neo-tree is on a left sidebar `require("neo-tree.sources.manager").get_state("filesystem", nil, winid).path` returns NIL, where `winid` refers to the correct neo-tree window. This should cover both cases, right?

**Reply by cseickel** [reply]
> More specifically, when neo-tree is on a left sidebar `require("neo-tree.sources.manager").get_state("filesystem", nil, winid).path` returns NIL, where `winid` refers to the correct neo-tree window. This should cover both cases, right?

No, it shouldn't. Maybe it would be cleaner if there were separate methods but I never really intended for that method to be used externally. There are 2 different types of trees:

1. Sidebar style, which is one per tab, identified by the tab number
2. Netrw style, which is one per window, identified by the window id

The `get_state()` function will return a tab state if the last argument is nil, or a window state if the last argument is not nil. In the case of a tab state, it is identified by the tab it belongs to and not the window. This roughly follows the convention of vim's `getcwd()`, albeit in reverse order. The reason for this is historic, in the beginning there was only sidebars and therefor all state was identified by the tab number.

If what you want is the state of a sidebar in the current tab, just use this: 
```lua
require("neo-tree.sources.manager").get_state("filesystem")
```

**Reply by wookayin** [reply]
`require("neo-tree.sources.manager").get_state("filesystem", nil, winid)` does not have `path` property where `winid` is the window id on the "netrw" style. When `winid = nil` (the sidebar style), path is retrieved correctly. Would this be a bug (v2.60)?

**Reply by cseickel** [reply]
@wookayin The path property will not be set until after the first navigation. My guess is that you are retrieving the state before the window has been shown, or you are calling the function with the wrong winid and getting a brand new state for the wrong window.
